### PR TITLE
Add a function to create CustomParticles from molecule symbols.

### PR DESCRIPTION
**Pull request by Quettle**
This new functions is useful to build objects representing molecules from user input or when parsing input files The mass of the created `CustomParticle` is simply the sum of the masses of its components. If the input is a simple particle, then a `Particle` object is created.

- [x] I have added a changelog entry for this pull request.

- [x] If adding new functionality, I have added tests and
      docstrings.

- [x] I have fixed any newly failing tests.

### CODE REVIEW

**Comment by Quettle** [comment]
> @Quettle — Thank you for working on this! This seems like it will be really helpful for the cold plasma community.
> 
> The main thing I'm wondering is whether or not it would be worthwhile to eventually create a `Molecule` class. For now I would be happy for PlasmaPy to include a `molecule` function that returns a `CustomParticle`, and then consider creating a `Molecule` class later on which would would use a lot of the same functionality.
> 
> I have a few suggestions for my first review, and I was also thinking that it would be helpful to have examples in the docstring for `molecule` to show how to use it.
> 
> Thank you again!

Thank you for your suggestions. I was also wondering if a Molecule class should be made before implementing a parser, but by just adding the masses of the element as I am doing here, I miss a lot of the precision found in the Particle class. CustomParticle is enough for the scope of my addition to the module. 
About the CustomParticle class, I will probably raise an issue to make it more convenient to work with when dealing with molecules. Since I want to be able differentiate isomers, the symbol is compared within the __eq__ method, but it doesn't appears in the __rep__.
One final thing, this implementation doesn't support symbols with repeating groups in parenthesis. Should I go the extra mile to make the parsing recursive in those cases?
Best regards,
Nicolas

**Comment by Quettle** [comment]
Hi, It seems that a major release has happened since I started this PR. the merging might get a bit messy. Should I start again from the current main branch?

**Comment by namurphy** [comment]
> Hi, It seems that a major release has happened since I started this PR. the merging might get a bit messy. Should I start again from the current main branch?

It doesn't look like there are any merge conflicts that have arisen between this PR and the `main` branch, so it would be fine to continue on this branch.  I've been doing some refactoring in `plasmapy.particles` lately, but I don't think my changes are going to interfere with this PR.  

Also, it looks like I started a review for this a while back but then forgot to click submit.  I'll make sure to finish that review today.  Sorry about that!

**Review by namurphy — CHANGES_REQUESTED** [review]
Thank you for submitting this!  It looks like a very useful function.  Most of my suggestions are pretty minor, with one question about making sure that this raises an exception for cases where the symbol is a valid `Particle` but not an element.  And my apologies for my delay in finishing this review.

I'm also wondering about possibilities to expand this functionality in the future, after this pull request is merged.  We could eventually create a `Molecule` class, though that'd probably be best done as a refactoring of `plasmapy.particles` to create a particle class hierarchy.  Another possibility to expand this functionality would be to allow molecule symbols that include Unicode symbols to get parsed.  That'd be something like...`molecule("H₂O")`.  It'd also be possible to create aliases for molecule names like we do for `Particle` [i.e., `molecule("water")`].  But, everything in this paragraph is beyond the scope of this pull request.

Thank you again!

File: plasmapy/particles/particle_class.py [hunk 1]
```diff
@@ -2216,6 +2218,82 @@ def __hash__(self) -> int:
...
+    `ParticleWarning`
+        If The charge is given both as an argument and in the symbol.
+
+    Examples
+    --------
+    >>> from plasmapy.particles import molecule
+    >>> molecule("I2")
+    CustomParticle(mass=4.214596603223354e-25 kg, charge=0.0 C)
+
+    Charge information is given either wint the symbol or as a second parameter.
```
[the remark this answers is not in the document]
Comment (Quettle) [reply plasmapy/particles/particle_class.py:2255]: Yes I missed the h of the with but I trust your English more than mine.

File: plasmapy/particles/particle_class.py [hunk 2]
```diff
@@ -2216,6 +2218,82 @@ def __hash__(self) -> int:
...
+    """
+    try:
+        return Particle(symbol, Z=Z)
+    except ParticleError:
+
+        element_dict, bare_symbol, Z = _parse_and_check_molecule_input(symbol, Z)
+        mass = 0 * u.kg
+        for element_symbol, amount in element_dict.items():
+            try:
+                element = Particle(element_symbol)
```
[the remark this answers is not in the document]
Comment (Quettle) [reply plasmapy/particles/particle_class.py:2285]: Indeed, I added the check for the element category.

File: plasmapy/particles/particle_class.py [hunk 3]
```diff
@@ -2216,6 +2218,89 @@ def __hash__(self) -> int:
...
+            except ParticleError as e:
+                raise InvalidParticleError(
+                    f"Could not identify {element_symbol}."
+                ) from e
+            if not element.is_category("element"):
+                raise InvalidParticleError(
+                    f"Molecule symbol contains a particle that is not an ellement :{element.symbol}"
```
Comment (namurphy) [inline plasmapy/particles/particle_class.py:2292]:
```suggestion
                    f"Molecule symbol contains a particle that is not an element: {element.symbol}"
```

File: plasmapy/particles/particle_class.py [hunk 4]
```diff
@@ -2216,6 +2218,89 @@ def __hash__(self) -> int:
...
+    us to differentiate between isomers:
+
+    >>> molecule("CH4O2") == molecule("CH3OOH")
+    False
+
+    """
+    try:
+        return Particle(symbol, Z=Z)
+    except ParticleError:
+
```
Comment (namurphy) [inline plasmapy/particles/particle_class.py:2280]:
```suggestion
```
[optional]

File: plasmapy/particles/particle_class.py [hunk 5]
```diff
@@ -2216,6 +2218,89 @@ def __hash__(self) -> int:
...
+    >>> molecule("Xe")
+    Particle("Xe")
+
+    The given symbol is preserved in the |CustomParticle| instance. This permits
+    us to differentiate between isomers:
+
+    >>> molecule("CH4O2") == molecule("CH3OOH")
+    False
+
```
Comment (namurphy) [inline plasmapy/particles/particle_class.py:2275]:
```suggestion
```
[optional]

File: plasmapy/particles/particle_class.py [hunk 6]
```diff
@@ -2216,6 +2218,89 @@ def __hash__(self) -> int:
...
+                    f"Molecule symbol contains a particle that is not an ellement :{element.symbol}"
+                )
+            mass += amount * element.mass
+
+        if Z is None:
+            charge = 0 * u.C
+        else:
+            charge = Z * const.e.si
+            bare_symbol += f" {-Z}-" if Z < 0 else f" {Z}+"
+        return CustomParticle(mass=mass, charge=charge, symbol=bare_symbol)
```
Comment (namurphy) [inline plasmapy/particles/particle_class.py:2301]:
```suggestion
            bare_symbol += f" {-Z}-" if Z < 0 else f" {Z}+"

        return CustomParticle(mass=mass, charge=charge, symbol=bare_symbol)
```
[optional]

**Review by namurphy — APPROVED** [review]
Thank you for making the changes so quickly!  It's looking just about ready. This time I have one suggested spelling change, as well as a few unimportant & optional stylistic suggestions that you're completely free to ignore.  Were there any other changes that you wanted to make, or should I go ahead and merge this?

**Comment by Quettle** [comment]
> Thank you for making the changes so quickly! It's looking just about ready. This time I have one suggested spelling change, as well as a few unimportant & optional stylistic suggestions that you're completely free to ignore. Were there any other changes that you wanted to make, or should I go ahead and merge this?

Thank you! I think It is complete. I wanted to make a small change on the CustomParticle to show the symbol in the __repr__ since it makes more of a difference now but I don't know if it deserves it's own PR since it may be breaking.

**Comment by namurphy** [comment]
> Thank you! I think It is complete. I wanted to make a small change on the CustomParticle to show the symbol in the **repr** since it makes more of a difference now but I don't know if it deserves it's own PR since it may be breaking.

Cool!  I'll go ahead and merge it then.  And good idea to isolate a breaking change to its own PR.  Showing the symbol in ``CustomParticle.__repr__`` would be a good idea now that you've made this change.  Thank you again!

**Pull request merged**